src/types/redis_cms.cc

```diff
@@ -0,0 +1,264 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#include "redis_cms.h"
+
+#include <stdint.h>
+
+#include "cms.h"
+#include "rocksdb/status.h"
+
+namespace redis {
+
+rocksdb::Status CMS::GetMetadata(engine::Context &ctx, const Slice &ns_key, CountMinSketchMetadata *metadata) {
+  return Database::GetMetadata(ctx, {kRedisCountMinSketch}, ns_key, metadata);
+}
+
+rocksdb::Status CMS::IncrBy(engine::Context &ctx, const Slice &user_key,
+                            const std::unordered_map<std::string, uint64_t> &elements) {
+  std::string ns_key = AppendNamespacePrefix(user_key);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  CountMinSketchMetadata metadata{};
+  rocksdb::Status s = GetMetadata(ctx, ns_key, &metadata);
+
+  if (s.IsNotFound()) {
+    return rocksdb::Status::NotFound();
+  }
+  if (!s.ok() && !s.IsNotFound()) {
+    return s;
+  }
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisCountMinSketch);
+  batch->PutLogData(log_data.Encode());
+
+  CMSketch cms(metadata.width, metadata.depth, metadata.counter, metadata.array);
+
+  if (elements.empty()) {
+    return rocksdb::Status::OK();
+  }
+
+  for (auto &element : elements) {
+    cms.IncrBy(element.first, element.second);
+    metadata.counter += element.second;
+  }
+
+  metadata.array = std::move(cms.GetArray());
+
+  std::string bytes;
+  metadata.Encode(&bytes);
+  batch->Put(metadata_cf_handle_, ns_key, bytes);
+
+  return storage_->Write(ctx, storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+}
+
+rocksdb::Status CMS::Info(engine::Context &ctx, const Slice &user_key, CMSketch::CMSInfo *ret) {
+  std::string ns_key = AppendNamespacePrefix(user_key);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  CountMinSketchMetadata metadata{};
+  rocksdb::Status s = GetMetadata(ctx, ns_key, &metadata);
+
+  if (!s.ok() || s.IsNotFound()) {
+    return rocksdb::Status::NotFound();
+  }
+
+  ret->width = metadata.width;
+  ret->depth = metadata.depth;
+  ret->count = metadata.counter;
+
+  return rocksdb::Status::OK();
+};
+
+rocksdb::Status CMS::InitByDim(engine::Context &ctx, const Slice &user_key, uint32_t width, uint32_t depth) {
+  std::string ns_key = AppendNamespacePrefix(user_key);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  CountMinSketchMetadata metadata{};
+
+  rocksdb::Status s = GetMetadata(ctx, ns_key, &metadata);
+
+  if (!s.IsNotFound()) {
+    return s;
```

Comment:
Shouldn't we return "key already exists" when `s.ok()`. If we follow the current way, it will return ok when init.